### DISCUSSION TITLE
Understanding clearsky and irradiance

### DISCUSSION THREAD

**Discussion by natecostello** [pvlib/pvlib-python#1961] [Q&A]
Way out of my depth here, so I apologize.

I'm struggling with how to interpret some data.  
Below is a plot of irradiance data generated by location.clearsky().  The plot origin is at sunrise, and its plotted through solar transit.  What I don't understand is why there is no irradiance of any flavor, including diffuse, until after sunrise.  My lived experience its that there is diffuse light before sunrise.  Am I missing parameters for atmospheric effects?

My ultimate interest is looking at edge cases for solar systems (sunrise/sunset + clouds) so any help in using pvlib to evaluate those cases would be greatly appreciated.

![image](https://github.com/pvlib/pvlib-python/assets/6625614/7368c36d-9bc5-4d46-839a-425d1e3ea32b)

Code behind that plot:
```python
hood_river = (45.71417626139969, -121.5154782530592, 'Hood River', 27, 'Etc/GMT+8')
latitude, longitude, name, altitude, timezone = hood_river

location = Location(
    latitude,
    longitude,
    name=name,
    altitude=altitude,
    tz=timezone,
    )

times = pd.date_range('2019-01-01 00:00', '2019-01-01 23:59', freq='1s', tz=location.tz)
weather = location.get_clearsky(times)
```

**Comment by cwhanse** [comment]
The function you are using (`get_clearsky` defaults to `irradiance.ineichen`) is coded to return ghi=0 when `apparent_zenith` > 90. That's probably why your calculation shows 0 diffuse. `apparent_zenith` is where the sun appears to be so it accounts for atmospheric refraction. I think you are also hoping for a model that predicts diffuse irradiance before apparent sunrise, during twilight periods. I'm not aware of a model designed to do that. Clearsky models tend to emphasize predictions with the sun well above the horizon.

**Reply by kandersolar** [reply]
The human eye is sensitive to light levels far below what is useful in PV contexts.  For example, moonlight is measured in mW/m^2, which eyes can readily make use of but PV systems famously not.  So the eye is not always a good judge of whether a light level is relevant for PV.

IMHO it's not too bad of an approximation to set the irradiance before/after sunrise/sunset to zero.  Here is a plot showing measurements of only ~5 W/m2 (0.5% of "full sun" irradiance) as the sun crosses the horizon:
![image](https://github.com/pvlib/pvlib-python/assets/57452607/d2fae2b5-1c04-4138-8faa-3e5bf535d7b6)

Source: https://midcdmz.nrel.gov/apps/plot.pl?site=BMS;start=20240101;edy=1;emo=2;eyr=2024;zenloc=222;amsloc=224;year=2024;month=1;day=30;time=1;inst=3;inst=130;type=plot;preset=0;first=3;math=0;second=-1;value=0.0;global=-1;direct=-1;diffuse=-1;user=0;axis=1

**Reply by natecostello** [reply]
Thanks for the replies.  That makes sense, especially now that I think about light levels and photography and how different the subjective is from the objective.

Can either of you recommend/identify a way to introduce cloud cover when it comes to clearsky?  Is that possible?

**Reply by kandersolar** [reply]
@natecostello Would your need be addressed by using pre-made weather/irradiance datasets that include the effect of cloud cover?  In fact most PV system modeling is done using weather datasets that model the effect of clouds on irradiance based on satellite imagery.  Producing those datasets requires significant expertise, but pvlib has functions to retrieve such weather datasets from external sources: https://pvlib-python.readthedocs.io/en/stable/user_guide/weather_data.html

**Reply by natecostello** [reply]
Yes...sort of.  The "sort of" caveat is because for what I'm trying to do, I need rather high temporal resolution, higher than is what is generally available from the weather/irradiance datasets.  Also since what I'm doing is comparing between system configurations, I really just need representative vice historical data.  The attraction of clearsky data is I can specify high temporal resolution.  But to accomodate bounding cases to compare between systems (and maintain high temporal resolution), I'd need to run clearsky data through some function that transforms it into cloudy-sky data.  I did get psm3 data working yesterday, which has resolution down to 5min, which is certainly a step in the right direction.  For context, here is an example of  what I'm trying to do (comparing ideal-efficiency MPPT output when MPPT voltage limits are in place vs PV production within 1 hour of sunrise - 1 sec temporal res).
<img width="622" alt="image" src="https://github.com/pvlib/pvlib-python/assets/6625614/bd841917-7c91-4108-9c40-1c8b9ae60edd">

In summary I'm seeking a function that could take clearsky as an input, and a cloud cover parameter or something, and output the resulting irradiance in a physically representative way.

**Reply by cwhanse** [reply]
If you can locate it, NREL published synthetic irradiance at 4s resolution about 10 years ago. Google term could be Solar Integration National Dataset (SIND), its chief developer was Marissa Hummon (no longer at NREL).
